pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -198,23 +219,152 @@ func (d *PmemCSIDriver) reconcileDeploymentChanges(r *ReconcileDeployment, chang
 		}
 	}
 
+	obsoleteList, err := d.getObsoleteObjects(r.operatorVersion)
+	if err != nil {
+		klog.Warningf("Failed to get obsolete objects for deployment %q: %v", d.Name, err)
+	}
+	for _, o := range obsoleteList {
+		if err := r.Delete(o); err != nil && !errors.IsNotFound(err) {
+			klog.Warningf("Failed to delete obsolete object %T for deployment %q: %v", o, d.GetName(), err)
+		}
+	}
+
+	if err := d.tagOwnedObjects(r.operatorVersion); err != nil {
+		klog.Warningf("Internal error while updating the annotation %s: %v", d.GetName(), err)
+	}
+
 	return false, nil
 }
 
-func (d *PmemCSIDriver) deployObjects(r *ReconcileDeployment) error {
-	objects, err := d.getDeploymentObjects()
-	if err != nil {
-		return err
+// tagOwnedObjects updates the deployment's annotation with the information(object name:GVK)
+// of the objects it owned as a json string.
+func (d *PmemCSIDriver) tagOwnedObjects(operatorVersion string) error {
+	nameGVKs := []string{}
+
+	// Nothing to update if this deployment is created by the same operator version
+	if d.ObjectMeta.Annotations[versionAnnotation] == operatorVersion {
+		return nil
 	}
-	for _, obj := range objects {
-		if err := r.Create(obj); err != nil {
+
+	for _, obj := range d.getDeploymentOwnedObjects() {
+		metaObj, err := meta.Accessor(obj)
+		if err != nil {
 			return err
 		}
+		gvk := obj.GetObjectKind().GroupVersionKind().String()
+		nameGVKs = append(nameGVKs, metaObj.GetName()+":"+gvk)
 	}
+
+	jsonString, err := json.Marshal(nameGVKs)
+	if err != nil {
+		return err
+	}
+
+	d.ObjectMeta.Annotations[versionAnnotation] = operatorVersion
+	d.ObjectMeta.Annotations[ownedObjectsAnnotation] = string(jsonString)
+	d.needsUpdate = true
+
 	return nil
 }
 
-// getDeploymentObjects returns all objects that are part of a driver deployment.
+// retrieveTaggedObjects retrives the objects tagged for this deployment by the
+// operator previously, on it's first successful reconcile
+func (d *PmemCSIDriver) retrieveTaggedObjects() ([]apiruntime.Object, error) {
+	objects := []apiruntime.Object{}
+	nameGVKs := []string{}
+
+	annotation, ok := d.ObjectMeta.Annotations[ownedObjectsAnnotation]
+	if !ok {
+		return objects, nil
+	}
+
+	if err := json.Unmarshal([]byte(annotation), &nameGVKs); err != nil {
+		return objects, err
+	}
+
+	for i := range nameGVKs {
+		parts := strings.SplitN(nameGVKs[i], ":", 2)
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("failed to parse annototion %q: %s", ownedObjectsAnnotation, nameGVKs[i])
+		}
+		gvk, err := parseGroupVersionKind(parts[1])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse annototion %q: %v", ownedObjectsAnnotation, err)
```

Comment:
typo: annototion -> annotation

Same above.

The annotation value is already in JSON format. Why don't you use JSON encoding also for the items inside the list? Then this entire code here goes away.

Reply:
Stored GVK is not a JSON string, its formatted by `GVK.String()`. So we cannot unmarshal that using JSON.

Reply:
And who says that we need to use GVK.String? Just define a list of structs which contain a GVK and you can marshal the entire list with json.